hail/python/hail/table.py

```diff
@@ -1168,6 +1168,109 @@ def write(self, output: str, overwrite = False, stage_locally: bool = False,
 
         Env.backend().execute(TableWrite(self._tir, output, overwrite, stage_locally, _codec_spec))
 
+    def _show_str(t, n, width, truncate, types):
+        width = max(width, 8)
+
+        if truncate:
+            truncate = min(max(truncate, 4), width - 4)
+        else:
+            truncate = width - 4
+
+        def trunc(s):
+            if len(s) > truncate:
+                return s[:truncate - 3] + "..."
+            else:
+                return s
+
+        def hl_repr(v):
+            if v.dtype == hl.tfloat32 or v.dtype == hl.tfloat64:
+                return hl.format('%.2e', v)
+            elif v.dtype == hl.tstr:
+                # FIXME doesn't quote
+                return hl.format('"%s"', v)
+            else:
+                return hl.str(v)
+
+        def hl_trunc(s):
+            return hl.cond(hl.len(s) > truncate,
+                           s[:truncate - 3] + "...",
+                           s)
+
+        def hl_format(v):
+            return hl.bind(lambda s: hl_trunc(s),
+                           hl.cond(hl.is_defined(v), hl_repr(v), "NA"))
+
+        t = t.flatten()
+        fields = [trunc(f) for f in t.row]
+        n_fields = len(fields)
+
+        types = [trunc(str(t.row[f].dtype)) for f in fields]
+        right_align = [hl.expr.types.is_numeric(t.row[f].dtype) for f in fields]
+
+        t = t.select(**{k: hl_format(v) for (k, v) in t.row.items()})
+        rows = t.take(n + 1)
+
+        has_more = len(rows) > n
+        rows = rows[:n]
+
+        rows = [[row[f] for f in fields] for row in rows]
+
+        column_width = [max(len(fields[i]), len(types[i]), max([len(row[i]) for row in rows]))
+                        for i in range(n_fields)]
+
+        column_blocks = []
+        start = 0
+        i = 1
+        w = column_width[0] + 4
+        while i < len(fields):
+            w = w + column_width[i] + 3
+            if w > width:
+                column_blocks.append((start, i))
+                start = i
+                w = column_width[i] + 4
+            i = i + 1
+        assert i == n_fields
```

Comment:
Do we have any plans to throw pretty errors?

Reply:
this is a true assertion - there's nothing the user can do that should trigger it.

Reply:
In that case, could it be placed in a unit test, to avoid runtime penalty (though small)?

Reply:
These kinds of assertions do ensure that our code doesn't enter a bad state, but also serve to provide some developer documentation. They're much better than written docs for enumerating invariants, because they're part of the code! Certainly worth the negligible performance hit.

They definitely don't take the place of unit tests, though.